### PR TITLE
Reordered anonymous type detection + cache

### LINKED ISSUE #447: Render Performance of Anonymous Types
### Dotliquid version
2.2.508

### Expected behavior
Anonymous types should be relatively similar in performance to explicitly converting to a Hash

### Actual behavior
A Hash built from a dictionary is significantly faster than a hash build from an anonymous type

### Steps to reproduce the Problem (you can add files)
Benchmarked the following object models:
```
            {Hash.FromAnonymousObject(new
            {
                products = Products
            });
```
vs
```
            Hash.FromAnonymousObject(new
            {
                products = Products.Select(product => new Hash()
                {
                    ["name"] = product.Name,
                    ["price"] = product.Price,
                    ["description"] = product.Description
                }).ToList()
            });
```
Where products is a simple list of a POCO with name, price, and description. Here is the performance from PR #441 
```
|              Method |        Job | BuildConfiguration | SyntaxCompatibilityLevel |      Mean |     Error |    StdDev | Ratio | RatioSD |   Gen 0 |  Gen 1 | Gen 2 | Allocated |
|-------------------- |----------- |------------------- |------------------------- |----------:|----------:|----------:|------:|--------:|--------:|-------:|------:|----------:|
|    RenderNestedHash | DefaultJob |            Default |              DotLiquid22 | 272.86 us |  1.418 us |  1.257 us |  0.55 |    0.01 | 46.3867 | 6.8359 |     - |    285 KB |
| RenderILiquidizable | DefaultJob |            Default |              DotLiquid22 | 578.03 us |  3.955 us |  3.506 us |  0.60 |    0.01 | 49.8047 | 5.8594 |     - |    309 KB |
```

### PR DESCRIPTION
**Pull request by microalps**
# Describe in your message, what you're fixing. Reference the original issue, if possible.
Implements a more performant way to detect anonymous types with caching. Resolves #447, or at least makes it much more acceptable. Dictionary/Hash is still faster than drops or anonymous types.

# Describe how if any API-breaking changes your PR may introduce.
No breaking changes

# State whether it's ready-to-review/-merge or if it's a WIP PR.
This is ready to merge for the issues addressed.

### CODE REVIEW

**Comment by microalps** [comment]
Performance comparison:
```
|              Method |        Job | BuildConfiguration | SyntaxCompatibilityLevel |      Mean |    Error |   StdDev |    Median | Ratio | RatioSD |   Gen 0 |  Gen 1 | Gen 2 | Allocated |
|-------------------- |----------- |------------------- |------------------------- |----------:|---------:|---------:|----------:|------:|--------:|--------:|-------:|------:|----------:|
|               Parse | Job-JMDNUK |              NuGet |              DotLiquid22 |  18.69 us | 0.341 us | 0.861 us |  18.32 us |  1.00 |    0.00 |  2.6245 | 0.0305 |     - |     16 KB |
|               Parse | DefaultJob |            Default |              DotLiquid22 |  18.60 us | 0.366 us | 0.612 us |  18.29 us |  0.97 |    0.04 |  2.6245 | 0.0305 |     - |     16 KB |
|                     |            |                    |                          |           |          |          |           |       |         |         |        |       |           |
|            ParseBig | Job-JMDNUK |              NuGet |              DotLiquid22 |  70.54 us | 0.960 us | 0.802 us |  70.50 us |  1.00 |    0.00 | 15.2588 | 1.0986 |     - |     94 KB |
|            ParseBig | DefaultJob |            Default |              DotLiquid22 |  69.06 us | 0.395 us | 0.369 us |  69.13 us |  0.98 |    0.01 | 15.2588 | 1.0986 |     - |     94 KB |
|                     |            |                    |                          |           |          |          |           |       |         |         |        |       |           |
|    RenderNestedHash | Job-JMDNUK |              NuGet |              DotLiquid22 | 283.51 us | 2.510 us | 2.789 us | 283.31 us |  1.00 |    0.00 | 43.9453 | 5.3711 |     - |    271 KB |
|    RenderNestedHash | DefaultJob |            Default |              DotLiquid22 | 267.96 us | 2.073 us | 1.731 us | 267.43 us |  0.94 |    0.01 | 43.9453 | 5.3711 |     - |    271 KB |
|                     |            |                    |                          |           |          |          |           |       |         |         |        |       |           |
| RenderILiquidizable | Job-JMDNUK |              NuGet |              DotLiquid22 | 586.84 us | 2.824 us | 2.358 us | 586.98 us |  1.00 |    0.00 | 49.8047 | 5.8594 |     - |    309 KB |
| RenderILiquidizable | DefaultJob |            Default |              DotLiquid22 | 337.23 us | 5.656 us | 7.929 us | 333.08 us |  0.58 |    0.01 | 49.8047 | 5.8594 |     - |    307 KB |
|                     |            |                    |                          |           |          |          |           |       |         |         |        |       |           |
|          RenderAnon | Job-JMDNUK |              NuGet |              DotLiquid22 | 759.73 us | 4.053 us | 3.791 us | 759.43 us |  1.00 |    0.00 | 52.7344 | 5.8594 |     - |    324 KB |
|          RenderAnon | DefaultJob |            Default |              DotLiquid22 | 329.97 us | 3.431 us | 3.209 us | 328.27 us |  0.43 |    0.00 | 48.3398 | 5.8594 |     - |    298 KB |
```

File: src/DotLiquid/Context.cs [hunk 1]
```diff
@@ -716,7 +704,11 @@ private static bool IsKeyValuePair(object obj)
             if (obj != null)
             {
                 Type valueType = obj.GetType();
```
Comment (microalps) [inline src/DotLiquid/Context.cs:706]: Null check is superfluous (line 650), as is GetType() call which is now assigned to objType on line 667.

File: src/DotLiquid/Context.cs [hunk 2]
```diff
@@ -524,14 +524,15 @@ private bool TryGetVariable(IEnumerator<string> partsEnumerator, out object vari
 
                 // If object is a KeyValuePair, we treat it a bit differently - we might be rendering
                 // an included template.
-                if (IsKeyValuePair(@object) && (part.SafeTypeInsensitiveEqual(0L) || part.Equals("Key")))
+                var isKvp = IsKeyValuePair(@object);
```
Comment (daviburg) [inline src/DotLiquid/Context.cs:527]:
```suggestion
                var isKeyValuePair = IsKeyValuePair(@object);
```
Avoid use of uncommon acronyms.

File: src/DotLiquid/Context.cs [hunk 3]
```diff
@@ -654,39 +655,26 @@ private static object Liquidize(object obj)
             {
                 return liquidizableObj.ToLiquid();
             }
-            if (obj is string)
+            if (obj is string || obj is IEnumerable || obj is decimal || obj is DateTime || obj is DateTimeOffset || obj is TimeSpan || obj is Guid)
             {
                 return obj;
             }
             if (obj is IEnumerable)
```
Comment (daviburg) [inline src/DotLiquid/Context.cs:662]: Remove this now, right?

File: src/DotLiquid/Util/ReflectionCache.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
+using System;
+using System.Collections.Generic;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+
+namespace DotLiquid.Util
+{
+    internal class ReflectionCache
```
Comment (daviburg) [inline src/DotLiquid/Util/ReflectionCache.cs:8]: nit: XML comments for the class, which is an opportunity to document for code maintainers why we have this cache class.

Comment (daviburg) [inline src/DotLiquid/Util/ReflectionCache.cs:8]:
```suggestion
    internal class ReflectionCacheValue
```
This is not the cache itself, but the class for a value within the cache.

File: src/DotLiquid/Util/ReflectionCache.cs [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    {
+        private const TypeAttributes AnonymousTypeAttributes = TypeAttributes.NotPublic;
+        private bool? _isAnonymous;
```
Comment (daviburg) [inline src/DotLiquid/Util/ReflectionCache.cs:11]: Can we use `Lazy` here instead of doing our own initialization sequence?

**Review by daviburg — CHANGES_REQUESTED** [review]

File: src/DotLiquid/Context.cs (hunk 3 above)
Comment (microalps) [reply src/DotLiquid/Context.cs:662]: No, we just combined a bunch of lines into one. But you are right `obj is IEnumerable` is duplicated.

**Review by daviburg — APPROVED** [review]

**Pull request merged**